File: exec/linux/compare/spot/tests/python/split.py
# -*- mode: python; coding: utf-8 -*-
# Copyright (C) by the Spot authors, see the AUTHORS file for details.
#
# This file is part of Spot, a model checking library.
#
# Spot is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# Spot is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
# License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import spot
import buddy
from unittest import TestCase
tc = TestCase()

# CPython use reference counting, so that automata are destructed
# when we expect them to be.   However other implementations like
# PyPy may call destructors latter, causing different output.
from platform import python_implementation
if python_implementation() == 'CPython':
    def gcollect():
        pass
else:
    import gc
    def gcollect():
        gc.collect()



def incl(a, b):
    return not b.intersects(spot.dualize(spot.tgba_determinize(a)))


def equiv(a, b):
    return incl(a, b) and incl(b, a)


def do_split(f, out_list):
    aut = spot.translate(f)
    outputs = buddy.bddtrue
    for a in out_list:
        outputs &= buddy.bdd_ithvar(aut.register_ap(a))
    s = spot.split_2step(aut, outputs, False)
    return aut, s

aut, s = do_split('(FG !a) <-> (GF b)', ['b'])
tc.assertTrue(equiv(aut, spot.unsplit_2step(s)))

del aut
del s
gcollect()

aut, s = do_split('GFa && GFb', ['b'])
tc.assertTrue(equiv(aut, spot.unsplit_2step(s)))
# FIXME s.to_str() is NOT the same on Debian stable and on Debian unstable
#       we should investigate this.  See Issue #502.
# tc.assertEqual("""HOA: v1
# States: 3
# Start: 0
# AP: 2 "a" "b"
# acc-name: generalized-Buchi 2
# Acceptance: 2 Inf(0)&Inf(1)
# properties: trans-labels explicit-labels trans-acc complete
# properties: deterministic
# spot-state-player: 0 1 1
# --BODY--
# State: 0
# [0] 1
# [!0] 2
# State: 1
# [!1] 0 {0}
# [1] 0 {0 1}
# State: 2
# [!1] 0
# [1] 0 {1}
# --END--""", s.to_str() )

del aut
del s
gcollect()

aut, s = do_split('! ((G (req -> (F ack))) && (G (go -> (F grant))))', ['ack'])
tc.assertTrue(equiv(aut, spot.unsplit_2step(s)))

# FIXME s.to_str() is NOT the same on Debian stable and on Debian unstable
#       we should investigate this.  See Issue #502.
# tc.assertEqual(s.to_str(), """HOA: v1
# States: 9
# Start: 0
# AP: 4 "ack" "req" "go" "grant"
# acc-name: Buchi
# Acceptance: 1 Inf(0)
# properties: trans-labels explicit-labels state-acc
# --BODY--
# State: 0
# [1&!2] 3
# [!1&!2] 4
# [1&2] 5
# [!1&2] 6
# State: 1
# [t] 7
# State: 2
# [t] 8
# State: 3
# [t] 0
# [!0] 1
# State: 4
# [t] 0
# State: 5
# [t] 0
# [!0] 1
# [!3] 2
# State: 6
# [t] 0
# [!3] 2
# State: 7 {0}
# [!0] 1
# State: 8 {0}
# [!3] 2
# --END--""")

del aut
del s
gcollect()

aut, s = do_split('((G (((! g_0) || (! g_1)) && ((r_0 && (X r_1)) -> (F (g_0 \
    && g_1))))) && (G (r_0 -> F g_0))) && (G (r_1 -> F g_1))',
                  ['g_0', 'g_1'])
tc.assertTrue(equiv(aut, spot.unsplit_2step(s)))

File: exec/linux/compare/spot/spot/twaalgos/split.cc
// -*- coding: utf-8 -*-
// Copyright (C) by the Spot authors, see the AUTHORS file for details.
//
// This file is part of Spot, a model checking library.
//
// Spot is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// Spot is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
// License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include "config.h"
#include <spot/twaalgos/split.hh>
#include <spot/misc/bddlt.hh>
#include <spot/priv/robin_hood.hh>
#include <unordered_set>

namespace spot
{
  namespace
  {
    template<typename genlabels>
    twa_graph_ptr split_edges_aux(const const_twa_graph_ptr& aut,
                                  genlabels gen)
    {
      twa_graph_ptr out = make_twa_graph(aut->get_dict());
      out->copy_acceptance_of(aut);
      out->copy_ap_of(aut);
      out->prop_copy(aut, twa::prop_set::all());
      out->new_states(aut->num_states());
      out->set_init_state(aut->get_init_state_number());

      // We use a cache to avoid the costly loop around minterms_of().
      // Cache entries have the form (id, [begin, end]) where id is the
      // number of a BDD that as been (or will be) split, and begin/end
      // denotes a range of existing transition numbers that cover the
      // split.
      //
      // std::pair causes some noexcept warnings when used in
      // robin_hood::unordered_map with GCC 9.4.  Use robin_hood::pair
      // instead.
      typedef robin_hood::pair<unsigned, unsigned> cached_t;
      robin_hood::unordered_map<int, cached_t> split_cond;

      internal::univ_dest_mapper<twa_graph::graph_t> uniq(out->get_graph());

      for (auto& e: aut->edges())
        {
          bdd cond = e.cond;
          if (cond == bddfalse)
            continue;
          unsigned dst = e.dst;
          if (aut->is_univ_dest(dst))
            {
              auto d = aut->univ_dests(dst);
              dst = uniq.new_univ_dests(d.begin(), d.end());
            }

          auto& [begin, end] = split_cond[cond.id()];
          if (begin == end)
            {
              begin = out->num_edges() + 1;
              for (bdd minterm: gen(cond))
                out->new_edge(e.src, dst, minterm, e.acc);
              end = out->num_edges() + 1;
            }
          else
            {
              auto& g = out->get_graph();
              for (unsigned i = begin; i < end; ++i)
                out->new_edge(e.src, dst, g.edge_storage(i).cond, e.acc);
            }
        }
      return out;
    }
  }

  twa_graph_ptr split_edges(const const_twa_graph_ptr& aut)
  {
    bdd all = aut->ap_vars();
    return split_edges_aux(aut, [&](bdd cond) {
      return minterms_of(cond, all);
    });
  }

  void edge_separator::add_to_basis(bdd label)
  {
    if (label == bddfalse)
      return;
    // Split our current set of labels using this new one.
    //
    // Do not use a range-based or iterator-based for loop here,
    // as push_back invalidates the end iterator.
    for (unsigned cur = 0, sz = basis_.size(); cur < sz; ++cur)
      if (bdd common = basis_[cur] & label;
          common != bddfalse && common != basis_[cur])
        {
          basis_[cur] -= label;
          basis_.push_back(common);
        }
  }

  void edge_separator::add_to_basis(const const_twa_graph_ptr& aut)
  {
    for (formula f: aut->ap())
      aps_.insert(f);

    robin_hood::unordered_set<int> seen{bddtrue.id()};
    for (auto& e: aut->edges())
      if (bdd lab = e.cond; seen.insert(lab.id()).second)
        add_to_basis(lab);
  }

  bool edge_separator::add_to_basis(const const_twa_graph_ptr& aut,
                                    unsigned long max_label)
  {
    std::set<bdd, bdd_less_than> all_labels;
    for (auto& e: aut->edges())
      {
        if (all_labels.insert(e.cond).second)
          if (max_label-- == 0)
            return false;
      }
    for (bdd lab: all_labels)
      add_to_basis(lab);
    for (formula f: aut->ap())
      aps_.insert(f);
    return true;
  }

  twa_graph_ptr
  edge_separator::separate_implying(const const_twa_graph_ptr& aut)
  {
    auto res = split_edges_aux(aut, [this](bdd cond) {
      return this->separate_implying(cond);
    });
    for (formula f: aps_)
      res->register_ap(f);
    return res;
  }

  twa_graph_ptr
  edge_separator::separate_compat(const const_twa_graph_ptr& aut)
  {
    auto res = split_edges_aux(aut, [this](bdd cond) {
      return this->separate_compat(cond);
    });
    for (formula f: aps_)
      res->register_ap(f);
    return res;
  }

  twa_graph_ptr separate_edges(const const_twa_graph_ptr& aut)
  {
    edge_separator es;
    es.add_to_basis(aut);
    return es.separate_implying(aut);
  }
}
